tools/hardware/utils/memory.py

```diff
@@ -26,6 +26,38 @@ def visitor(node: WrappedNode):
     return regions
 
 
+def merge_memory_regions(regions: Set[Region]) -> Set[Region]:
+    ''' Check all region and merge adjacent ones '''
+    all_regions = []
+    idx = 0
+    for region in regions:
+        all_regions.append(dict(idx=idx, region=region, right_adj=None, left_adj=None))
+        idx += 1
+
+    # Find all right contiguous regions
+    for dreg in all_regions:
+        for dnreg in all_regions[dreg['idx']+1:]:
+            if dreg['region'].owner == dnreg['region'].owner:
+                if dnreg['region'].base == dreg['region'].base + dreg['region'].size:
+                    dreg['right_adj'] = dnreg
+                    dnreg['left_adj'] = dreg
+                elif dreg['region'].base == dnreg['region'].base + dnreg['region'].size:
+                    dnreg['right_adj'] = dreg
+                    dreg['left_adj'] = dnreg
+
+    # Find all the left-most contiguous regions
+    contiguous_regions = set()
+    for reg in all_regions:
+        if reg['left_adj'] is None:
+            size = reg['region'].size
+            r_adj = reg['right_adj']
+            while r_adj is not None:
+                size += r_adj['region'].size
+                r_adj = r_adj['right_adj']
+            contiguous_regions.add(Region(reg['region'].base, size, reg['region'].owner))
+    return contiguous_regions
```

Comment:
please add an empty line here, that improves code readability a bit

Reply:
Done.